strawberryfields/apps/qchem/dynamics.py

```diff
@@ -0,0 +1,108 @@
+main parts. First, preparation of a vibrationally excited initial state. Second, application of the
+dynamics transformation :math:`U(t)` and finally, generating samples and computing the probability
+of observing desired states. It is noted that the initial states can be prepared in different ways.
+For instance, they can be Fock states or Gaussian states such as coherent or two-mode squeezed
+vacuum states.
+
+Algorithm
+---------
+
+The algorithm for simulating the vibrational quantum dynamics in the localized basis with a photonic
+device has the following form:
+
+1. Each initial optical mode is assigned to a vibrational local mode and a specific excitation is
+   created using one of the state preparation methods discussed.
+
+2. An interferometer is configured according to the unitary :math:`U_l^\dagger` and the initial
+   state is propagated through the interferometer.
+
+3. For each mode, a rotation gate is designed according to the vibrational frequency of the
+   corresponding normal mode (:math:`\nu`) as
+   :math:`R(\theta) = \exp(i\theta \hat{a}^{\dagger}\hat{a})` where :math:`\theta = -2 \pi \nu t`
+   and :math:`t` is time.
+
+4. A second interferometer is configured according to the unitary :math:`U_l` and the new state
+   is propagated through the interferometer.
+
+5. The number of photons in each output mode is measured.
+
+6. Samples are generated and the probability of obtaining a specific excitation in a given mode
+   (or modes) is computed for time :math:`t`.
+
+This module contains functions for implementing this algorithm. The function `dynamics_observable`
+return a custom ``sf`` operation that contains the required unitary and rotation operations
+explained in steps 2-4 of the algorithm.
+"""
+from typing import Callable
+
+from scipy.constants import c, pi
+
+import numpy as np
+
+import strawberryfields as sf
+from strawberryfields.utils import operation
+
+
+def dynamics_observable(t: float, Ul: np.ndarray, w: np.ndarray) -> Callable:
+    r"""Generates a custom ``sf`` operation for performing the transformation
+    :math:`U(t) = U_l^\dagger e^{-i\hat{H}t/\hbar} U_l` on a given state.
+
+    **Example usage:**
+
+    >>> t = 0.0  # time
+    >>> Ul = np.array([[0.707106781, -0.707106781],
+    >>>                [0.707106781, 0.707106781]])  # normal to local transformation
+    >>> w = np.array([3914.92, 3787.59])  # frequencies
+    >>> obs =  dynamics_observable(t, Ul, w)
+
+    Args:
+        t (float): time (femtosecond)
+        Ul (array): normal to local transformation matrix
+        w (array): normal mode frequencies (:math:`\mbox{cm}^{-1}`)
+
+    Returns:
+        op (Callable): custom ``sf`` operation function
+    """
+    modes = len(Ul)
+    theta = -w * 100.0 * c * t * 1e-15 * (2 * pi)
+
+    # pylint: disable=expression-not-assigned,pointless-statement
+    @operation(modes)
+    def op(q):
+        sf.ops.Interferometer(Ul.T) | q
```

Comment:
Do we need to take the conjugate to get the dagger version?

Reply:
This matrix is always real. The normal to local transformation done by Ul is a purely classical process. Does it answer the question?

Reply:
Will this be clear to users, is it worth either checking if input unitaries are real or mentioning in the docstrings?

Reply:
I will add a condition to check that the unitary is real.